lib/types/addresstrie.go

```diff
@@ -0,0 +1,120 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	"net"
+	"strconv"
+	"strings"
+)
+
+// NullAddressTrie is a wrapper around AddressTrie like guregu/null
+type NullAddressTrie struct {
+	Trie  *AddressTrie
+	Valid bool
+}
+
+// NewNullAddressTrie returns valid (Valid: true) AddressTrie
+func NewNullAddressTrie(source map[string]HostAddress) NullAddressTrie {
+	return NullAddressTrie{
+		Trie:  NewAddressTrie(source),
+		Valid: true,
+	}
+}
+
+// MarshalJSON converts NullAddressTrie to valid JSON
+func (n NullAddressTrie) MarshalJSON() ([]byte, error) {
+	if !n.Valid { // envconfig bug?
+		return []byte("null"), nil
+	}
+
+	jsonMap := make(map[string]string)
+	for k, v := range n.Trie.source {
+		var right string
+		if v.Port != 0 {
+			right = v.String()
+		} else {
+			right = v.IP.String()
+		}
+		jsonMap[k] = right
+	}
+
+	return json.Marshal(jsonMap)
+}
+
+// UnmarshalJSON converts JSON to NullAddressTrie
+func (n *NullAddressTrie) UnmarshalJSON(data []byte) error {
+	if bytes.Equal(data, []byte("null")) {
```

Comment:
I would prefer it if we avoided using static strings as such in the code. They can be confusing as one would hardly know if they're supposed to be a "standardized" value or an arbitrary one. In this case, it would be worth declaring a constant string `null` or something along those lines.